### PR TITLE
Improve booking seats changes

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR continues #5427 by updating calendar & video events whenever there are changes in seats bookings. This abstracts some of the seats logic to a `handleSeats` function.

Fixes # (issue)

**Environment**: Staging(main branch) / Production

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How should this be tested?

- Create a booking with seats and add a destination calendar
- Fill the booking with attendees
- As an attendee
    - Move to a different time slot that has no booking
        - There should be a new booking created and the destination calendar should be updated with the new calendar event and the previous one should have that attendee removed
    - For that new booking you should be the only attendee. Reschedule to another empty time slot
        - On the calendar, that previous booking should be removed and a new one created
    - Reschedule back to the booking that has already filled with seats
        - The calendar event should now include the new attendee
    - Now cancel the event as the attendee
        - The calendar event should still be there but now without the attendee
- As the owner
    - Reschedule the booking
        - The whole booking should have been moved
    -  Cancel the booking
        - The whole booking should have been canceled

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 1]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
+
+    if (videoCallReference) {
+      evt.videoCallData = {
+        type: videoCallReference.type,
+        id: videoCallReference.meetingId,
+        password: videoCallReference?.meetingPassword,
+        url: videoCallReference.meetingUrl,
+      };
+    }
+  };
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:668]: If we're updating the original rescheduled booking calendar event we need to pass the video call data if there is any.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 2]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
+
+  /* Check if the original booking has no more attendees, if so delete the booking 
+  and any calendar or video integrations */
+  const lastAttendeeDeleteBooking = async (
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:672]: If a booking is reaching 0 attendees, then we should delete the booking from the DB and delete any calendar / video integrations

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 3]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
+        }
+      }
+
+      await Promise.all(integrationsToDelete).then(async () => {
+        await prisma.booking.delete({
+          where: {
+            id: originalRescheduledBooking.id,
+          },
+        });
+      });
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:713]: After we make the calls to the integrations then we can delete the booking from the DB

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
       throw new HttpError({ statusCode: 404, message: "Booking not found" });
     }
 
-    // Need to add translation for attendees to pass type checks. Since these values are never written to the db we can just use the new attendee language
-    const bookingAttendees = booking.attendees.map((attendee) => {
-      return { ...attendee, language: { translate: tAttendees, locale: language ?? "en" } };
-    });
+    if (dayjs(booking.startTime).utc().format() === evt.startTime) {
+      return booking;
+    }
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:744]: If we are rescheduling and the same timeslot is chosen then just return the booking

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 5]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
+      }
 
-    if (videoCallReference) {
-      evt.videoCallData = {
-        type: videoCallReference.type,
-        id: videoCallReference.meetingId,
-        password: videoCallReference?.meetingPassword,
-        url: videoCallReference.meetingUrl,
-      };
-    }
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:637]: Abstracted this to a function this is needed in multiple places

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 6]
```diff
@@ -586,153 +637,525 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
         },
-        booking: {
-          connect: {
+      });
+
+      const credentials = await refreshCredentials(organizerUser.credentials);
+      const eventManager = new EventManager({ ...organizerUser, credentials });
+
+      // This is passed to the event manager to update the original calendar event
+      let originalBookingEvt;
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:782]: Create a `CalendarEvent` object from the original rescheduled booking. We need this to pass to the event manager to update the original calendar / video integrations

File: packages/lib/CalEventParser.ts [hunk 7]
```diff
@@ -137,6 +137,10 @@ export const getUid = (calEvent: CalendarEvent): string => {
   return uid ?? translator.fromUUID(uuidv5(JSON.stringify(calEvent), uuidv5.URL));
 };
 
+const getSeatReferenceId = (calEvent: CalendarEvent): string => {
+  return calEvent.attendeeSeatId ? `seatReferenceUId=${calEvent.attendeeSeatId}` : "";
+};
+
```
Comment (joeauyeung) [inline packages/lib/CalEventParser.ts:143]: Pass the seat referenceUId to the email reschedule and cancel links incase the attendee does not have a cal account

**Review by zomars — APPROVED** [review]
Merging to the other PR to consolidate and prevent more conflicts. Let's finish it on the other PR.

**Pull request merged**